Fixes issue when using an ampersand (&) to scope css to a parent selector alongside the new namespace feature

marmite22 opened the pull request:
The instructions on raising an issue said to consider making a PR with a failing test so here is one:

The issue this tests for is that when using an ampersand (&) to scope css to a parent selector and also using the new namespace feature added in the latest beta version, the resulting css has the parent selector before the namespace selector.

I took a stab at resolving this but while I was able to get the original issue solved, it breaks down with more complex use cases (as seen in the 2nd test)

marmite22 commented:
Is anyone (@probablyup ?) able to verify the output of my 2nd test please? I've, perhaps wrongly, assumed that the outputted CSS should target the sc-* class but if I replicate the same code I get different results in v4 vs. v5. In fact, v5 actually switches between the sc-* class and the other generated class.

Given the following styled component
```
const StyledComponent = styled.div`
  color: blue;
  .child &,
  .child2 & {
    color: green;
  }
`;
```

### v4.4.1

```
.gMjrFY{color:blue;}
.child .gMjrFY,.child2 .gMjrFY{color:green;}
```
### v5.3.6

```
.QcRiW{color:blue;}
.child .sc-bcXHqe,.child2 .QcRiW{color:green;}
```

So why does v5 target the sc-* class where v4 didn't? Is it correct? If it is correct then my fix for the issue is working and it is the test that is wrong.

marmite22 commented:
I can spend some time in my next sprint at work to look at this but I need to know the answer to my question above about the sc- classes. 

I'll happily work more on this if anyone has any feedback?

probablyup commented:
The relevant documentation is here: https://styled-components.com/docs/basics#pseudoelements-pseudoselectors-and-nesting

marmite22 commented:
Looking at that documentation
![image](https://user-images.githubusercontent.com/2477900/213152563-d6248235-c5fd-4f9b-a9ab-84a432d067fb.png)
I can see the same issue is present (the sc-* class is being targeted instead of the other one).

If you are happy that the generated css targets those classes then I will update the test to pass and this can then be merged I think.

In fact, I think I will raise a completely separate issue specifically about it.

marmite22 commented:
I have now fixed the tests I've added so they are passing and raised a separate issue here https://github.com/styled-components/styled-components/issues/3923

I think this PR now resolves https://github.com/styled-components/styled-components/issues/3909

probablyup commented on packages/styled-components/src/utils/stylis.ts:
```diff
@@ -91,17 +91,22 @@ export default function createStylisInstance(
+        return rule;
+      });
+    }
+    const serialized = serialize(compiled, middleware(middlewares));
```
Can you drop this variable that isn’t reused? Looks good otherwise.

marmite22 replied:
done

probablyup commented on packages/styled-components/src/utils/stylis.ts:
```diff
@@ -91,17 +91,22 @@ export default function createStylisInstance(
+    if (options.namespace) {
+      compiled = compiled.map(rule => {
+        // add the namespace to the start
+        rule.value = `${options.namespace} ${rule.value}`;
```
Actually does this work appropriately with media queries? I’m not sure if we have a test for this explicitly

marmite22 replied:
I have added an extra test for a media query inside a namespaced selector. I had to write a recursive function to go down through the rules and ensure that the namespace was added inside the media query rather than before it. 

I think it works now!

probablyup reviewed (approved):
Thanks, looks good!
